docs/arch/002-store-responses-in-query-string.md

```diff
@@ -0,0 +1,61 @@
+
+1. It is difficult to generate URLs with pre-populate responses, for example `/flow-name/y/response1/response2/<pre-populated-response>`. Unless the response is for the first question in a flow, it requires pre-filling responses for all former questions which are unlikely to be known. 
+
+1. Changes to a flow can cause issues with previously cached pages. As URLs are not deterministic for a particular page, changes to a flow can alter which page is displayed. For example, a given flow displays an outcome page for the path `/x/y/z`. A question is removed and the flow now shows an outcome page for the path `/x/y`. The cache would continue to return the old question and users would then be redirected to `/x/y/z`, which is no longer a valid path and could result in an error if it was not also cached.
+
+1. We lose existing responses if the user visits an earlier question. To view a previous or earlier question, responses need to be removed from the URL. This means that it is impossible to change a previous response without re-answering all subsequent questions. This can be a particular burden for users if the flow is long to complete, requires complex input (e.g. dates) or information that isn't memorised.
+
+Additionally, the implementation of storing user responses in a session required the creation of a seperate controller `SessionAnswersController`, separate logic for node resolution and updating the flow. This adds extra burden when maintaining or improving the codebase, as new code may need to accommodate both response storage methods.
+
+## Proposal
+
+Add the ability to store user responses using the URL query string. This will become the default method for storing responses and replace use of the URL path. The ability to store responses in a session cookie will remain and require flow to explicitly enable it within configuration (current behaviour).
+
+To implement this change, we could leverage the existing code (controller, node resolution and flow updates) written for storing responses in a session. This is because the underlying data structure storing the responses is the same i.e. a hash keyed by node name. The only additional logic required is to retrieve and update the parameters in the query string instead from the session.
+
+Example:
+- `flow-name/y` => `flow-name/s/question1`
+- `flow-name/y/response1` => `flow-name/s/question2?question1=response1`
+- `flow-name/y/response1/response2` => `flow-name/s/results?question1=response1&question2=question2`
+
+Suggested steps for implementation:
+
+1. Rename `SessionAnswersController` to `FlowController` (as will be used to handle all flows)
+1. Change `use_session` configuration option in flow definition to specify different response store types. e.g. `response_store :session`
+1. Add functionality to store responses in query string
+1. Migrate existing flows incrementally to use query string by specifying config option e.g. `response_store :query_string`
```

Comment:
Good catch. We need to cover those who are in the middle of doing a flow, and old links that might be saved.

Feels like there's scale of effort vs disruption for users and depends on the amount of traffic (how many user's would be effected).

Most effort, minimal disruption:
Add logic to translate old urls into new query string based ones.

Medium effort, medium disruption:
Redirect all old URLs to the start page.

Low effort, high disruption:
Drop the old routes for old URLs and 404.

Safest approach would be to:
1. Initially keep the existing controller to handle old URL requests (we can distinguish them by the extra `/y/` vs `/s/` segment after the flow name).
1. Add routes for both URL types in the content item (for router).
2. Update the start page to point to the query string based flow.
3. Keep an eye on traffic to the old URLs
4. Then decide on level of effort thats needed?

My gut feeling is we don't actually have that many people save URLs or link directly to results, and that we'd probably be fine to just redirect to the start page.

Reply:
Yeah I think it's fine to have something that is intermediary and then remove. I think that could be neater too as we could set redirects in Publishing API for all the /y URL's at the point of removal and then not need any /y handling code in this app.

How do you feel about using this as a chance to change /s/ and /y/ to something more understandable? a quick idea I had is /flow/ ? Just a thought given it's going to be a bit disruptive anyway.

Reply:
Yup agree - should try and add the redirect in Publishing API. Would let us really clean everything up.

I think that's a good idea about renaming /s/ - I'm slightly worried about adding extra work to implement this change - (would have to migrating the existing session based answer (find-coronavirus-support) as well). I also wonder if it's worth, outside this ADR, thinking about whether we can remove it all together.

Reply:
Great re: redirect

> I think that's a good idea about renaming /s/ - I'm slightly worried about adding extra work to implement this change - (would have to migrating the existing session based answer (find-coronavirus-support) as well). I also wonder if it's worth, outside this ADR, thinking about whether we can remove it all together.

Yeah getting rid altogether sounds ideal but I agree that's out of the scope of this.

I think the renaming should be in the scope of this as /s/ will leave a bit of a smell behind. The s stands for session as I recall? Since every other Smart Answer needs migrating anyway it'd seem the extra work is very low since it'd be a different application of the same process. Only tricky thing seems to be agreeing on what the rename should be 😁

Reply:
Sure, you are right it'll less work now than later. /flow/ does make sense - it's short and descriptive. Struggling to think of any alternatives,  maybe `/nodes/` (but prefer `flow`)?

Another thing is to make sure it's not going to confuse users who won't be familiar with these concepts.

Reply:
Sounds good - yeah it seems a step up from /y/ anyway for confusion.

Reply:
I've added a section talking about the prerequisite work to update /s/ to /flow/ - including suggested step of implementation.